Python/ceval.c

```diff
@@ -4603,16 +4635,44 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
 
         TARGET(CALL_FUNCTION): {
             PREDICTED(CALL_FUNCTION);
-            PyObject **sp, *res;
-            sp = stack_pointer;
-            res = call_function(tstate, &sp, oparg, NULL, cframe.use_tracing);
-            stack_pointer = sp;
-            PUSH(res);
-            if (res == NULL) {
+            PyObject *res;
+
+            // Check if the call can be inlined or not
+            PyObject *function = *(stack_pointer - oparg - 1);
+            int inline_call = 0;
+            if (Py_TYPE(function) == &PyFunction_Type) {
+                PyCodeObject *code = (PyCodeObject*)((PyFunctionObject*)(function))->func_code;
+                int is_coro = code->co_flags & (CO_GENERATOR | CO_COROUTINE | CO_ASYNC_GENERATOR);
+                inline_call = (is_coro || cframe.use_tracing) ? 0 : 1;
+            }
+
+            if (!inline_call) {
+                PyObject **sp = stack_pointer;
+                res = call_function(tstate, &sp, oparg, NULL, cframe.use_tracing);
+                stack_pointer = sp;
+                PUSH(res);
+                if (res == NULL) {
+                    goto error;
+                }
+                CHECK_EVAL_BREAKER();
+                DISPATCH();
+            }
+
+            assert(!cframe.use_tracing);
+            InterpreterFrame *new_frame = _PyEval_FrameFromPyFunctionAndArgs(tstate, stack_pointer-oparg, oparg, function);
+            if (new_frame == NULL) {
+                // When we exit here, we own all variables in the stack (the frame creation has not stolen
+                // any variable) so we need to clean the whole stack (done in the "error" label).
                 goto error;
             }
-            CHECK_EVAL_BREAKER();
-            DISPATCH();
+            assert(tstate->interp->eval_frame != NULL);
+            // The frame has stolen all the arguments from the stack, so there is no need to clean up them
+            STACK_SHRINK(oparg + 1);
+            Py_DECREF(function);
+            _PyFrame_SetStackPointer(frame, stack_pointer);
+            new_frame->depth = frame->depth + 1;
+            tstate->frame = frame = new_frame;
```

Comment:
Note to self, no need to pop frame here because it's done on eval frame exit at ``exit_eval_frame``.